rfc/rfc-66/rfc-66.md

```diff
@@ -0,0 +1,318 @@
+| Log file format            | Adding deltacommit instant time to log file name. The file name does not contain base instant time; No changes to log block format |
+
+### Commit Protocol
+
+| Type                            | Content                                                                                                                                       |
+|---------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------|
+| Writer expectations             | Writer does not need to look up the latest file slice to fetch the base commit time anymore, it always uses the current instant time directly |
+| Conflict Resolution             | No explicit conflict resolution, it is handled by the compactor                                                                               |
+| Reader expectations             | Readers still query based on file slices, see the pseudocode for new file slicing                                                             |
+| Synchronization and correctness | Still need some mutex on commit log (timeline). Getting the completion time and adding to the timeline needs to happen atomically             |
+
+**Note:** Clustering and updates conflict resolution is not in the scope of non-blocking concurrency control, you still need the OCC for course-grained failures and retries. 
+
+## Appendix
+
+### Pseudocode for New File Slicing
+
+Assume we have the following inputs to build the file slices:
+1. `base_files_by_file_id`: a map of filegroup id to list of base files.
+2. `log_files_by_file_id`: a map of feilgroup id to list of log files.
+3. `timeline`: Hudi write timeline .
+
+The pseudocode below introduces a notion of **file slice barriers**, which contain a list of instant (start) times of 
+the base files in descending order. Barriers will help in demarcating file slices. The below pseudocode builds the file 
+slices per file group. Building file groups from file slices is not shown but can be easily done.
+
+```python
+# new file slicing
+def build_file_slices(base_files_by_file_id, log_files_by_file_id, timeline):
+    # get set of all filegroup ids
+    file_id_set = base_files_by_file_id.keys
+    file_id_set.add_all(log_files_by_file_id.keys)
+    
+    for file_id in file_id_set:
+        # sort the base files by descending order of instant (start) time, i.e. last written base file first
+        base_files_in_file_id = base_files_by_file_id[file_id].sort(BASE_FILE_REVERSE_COMPARATOR)
+        # filter out log files that have been log-compacted
+        log_files_in_file_id = handle_log_compaction(log_files_by_file_id[file_id], timeline)
+        # sort the log files by ascending order of completion time
+        log_files_in_file_id.sort(LOG_FILE_COMPARATOR)
```

Comment:
It says sort by completion time here, meaning the deltacommit's completion time? But in the log file sorting rule section, it shows order by delta commit time, which I think you meant instant start time. So which one it is?